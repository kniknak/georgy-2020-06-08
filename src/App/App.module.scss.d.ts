declare const styles: {
    readonly root: string;
    readonly message: string;
    readonly error: string;
};
export = styles;