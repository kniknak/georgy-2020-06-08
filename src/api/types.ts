export interface FileInfo {
    originalName: string
    size: number
    _id: string
    createdAt: number
}

export enum Status {
    Initial = 1,
    Loading,
    Success,
    Error,
}