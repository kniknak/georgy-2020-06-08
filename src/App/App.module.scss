.root {
    max-width: 960px;
    margin: 0 auto;
}

.message {
    padding: 12px;
}

.error {
    padding: 12px;
    color: darkred;
}

