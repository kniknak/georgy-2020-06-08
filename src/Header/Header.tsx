import React from "react";
import { FileUpload } from "./FileUpload/FileUpload";
import { Search } from "./Search/Search";
import * as styles from "./Header.module.scss";

export interface HeaderProps {
    addFile: (file: File) => void
    search: string
    setSearch: (search: string) => void
}

export const Header: React.FC<HeaderProps> = ({ addFile, search, setSearch }) => {
    return (
        <div className={styles.root}>
            <div className={styles.search}>
                <Search
                    search={search}
                    setSearch={setSearch}
                />
            </div>
            <div className={styles.upload}>
                <FileUpload addFile={addFile}/>
            </div>
        </div>
    );
};
