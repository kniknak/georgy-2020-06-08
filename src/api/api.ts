import { FileInfo } from "./types";

const endpoint = "http://localhost:3003/files";

const validateResponse = async (res: Response) => {
    const data = await res.json();

    if (!res.ok || !data.status) {
        throw data;
    }

    return data;
};

export const apiFetchFiles = (search: string): Promise<{ data: FileInfo[], search: string }> => {
    const url = new URL(endpoint);

    if (search) {
        url.searchParams.append("search", search);
    }

    return fetch(url.href, { mode: "cors" })
        .then(validateResponse)
        .then(response => ({
            ...response,
            search,
        }));
};

export const apiAddFile = (file: File) => {
    if (file.size >= 10 * 1024 * 1024) {
        return Promise.reject(new Error("File size exceeds 10Mb"));
    }

    const data = new FormData();
    data.append("upload", file);

    return fetch(endpoint, {
        method: "POST",
        body: data,
        mode: "cors",
    })
        .then(validateResponse);
};

export const apiDeleteFile = (id: string) => (
    fetch(`${endpoint}/${id}`, {
        method: "DELETE",
        mode: "cors",
    })
        .then(validateResponse)
);