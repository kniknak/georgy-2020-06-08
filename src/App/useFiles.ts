import { useCallback, useState } from "react";
import { FileInfo, Status } from "../api/types";
import { apiAddFile, apiDeleteFile, apiFetchFiles } from "../api/api";
import throttle from "lodash.throttle";

export const useFiles = () => {
    const [status, setStatus] = useState(Status.Initial);
    const [message, setMessage] = useState("");
    const [searchValue, setSearchValue] = useState("");
    const [files, setFiles] = useState<FileInfo[]>([]);

    const onError = useCallback(error => {
        console.error(error);
        setFiles([]);
        setStatus(Status.Error);
        setMessage(error.message);
    }, [setStatus, setFiles]);

    const getFiles = useCallback(throttle((search: string) => {
        if (search && !search.match(/^[0-9a-zA-Z.]([\s0-9a-zA-Z.]*[0-9a-zA-Z.])?$/)) {
            return;
        }

        setStatus(Status.Loading);

        apiFetchFiles(search)
            .then(result => {
                if (result.search === search) {
                    setFiles(result.data);
                    setStatus(Status.Success);
                }
            })
            .catch(onError);
    }, 500), [setFiles, setStatus, onError, throttle]);

    const deleteFile = useCallback((id: string) => {
        apiDeleteFile(id)
            .then(() => getFiles(searchValue))
            .catch(onError);
    }, [getFiles, onError, searchValue]);

    const addFile = useCallback((file: File) => {
        apiAddFile(file)
            .then(() => getFiles(searchValue))
            .catch(onError);

    }, [getFiles, onError, searchValue]);

    const setSearch = useCallback((search: string) => {
        setSearchValue(search);
        return getFiles(search);
    }, [setSearchValue, getFiles]);

    return {
        status,
        files,
        getFiles,
        deleteFile,
        addFile,
        search: searchValue,
        setSearch,
        message,
    };
};