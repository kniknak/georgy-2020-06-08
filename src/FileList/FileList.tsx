import React from "react";
import { FileInfo } from "../api/types";
import { FileItem } from "./FileItem/FileItem";
import * as styles from "./FileList.module.scss";
import { FileListInfo } from "./FileListInfo/FileListInfo";

export interface FileListProps {
    files: FileInfo[]
    deleteFile: (id: string) => void
}

export const FileList: React.FC<FileListProps> = ({ files, deleteFile }) => {
    return (
        <div>
            <FileListInfo files={files}/>
            <ul className={styles.list}>
                {files.map(file => (
                    <FileItem
                        key={file._id}
                        file={file}
                        deleteFile={deleteFile}
                    />
                ))}
            </ul>
        </div>
    );
};
