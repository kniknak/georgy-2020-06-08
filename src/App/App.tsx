import React, { useEffect } from "react";
import { Header } from "../Header/Header";
import { FileList } from "../FileList/FileList";
import { useFiles } from "./useFiles";
import { Status } from "../api/types";
import * as styles from "./App.module.scss";

export const App: React.FC = () => {
    const {
        files,
        status,
        getFiles,
        deleteFile,
        addFile,
        search,
        setSearch,
        message
    } = useFiles();

    useEffect(() => getFiles(""), [getFiles]);

    const isLoading = status === Status.Loading;
    const hasError = status === Status.Error;

    return (
        <div className={styles.root}>
            <Header
                addFile={addFile}
                search={search}
                setSearch={setSearch}
            />
            {isLoading && <div className={styles.message} data-loading>Loading...</div>}
            {hasError && (
                <div className={styles.error} data-error>
                    Unexpected Error. {message || "Try again or reload the page"}
                </div>
            )}
            {!isLoading && !hasError && (
                <FileList
                    files={files}
                    deleteFile={deleteFile}
                />
            )}
        </div>
    );
};