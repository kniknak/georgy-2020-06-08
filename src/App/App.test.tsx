import React from "react";
import { shallow } from "enzyme";
import { useFiles } from "./useFiles";
import { FileInfo, Status } from "../api/types";
import { App } from "./App";
import { Header } from "../Header/Header";
import * as styles from "./App.module.scss";
import { FileList } from "../FileList/FileList";

jest.mock("./useFiles");
jest.mock("react", () => ({
    ...jest.requireActual("react"),
    useEffect: (f: Function) => f(),
}));

const mockedUseFiles = useFiles as jest.Mock<ReturnType<typeof useFiles>>;

describe("App", () => {
    const defaultUseFilesMockReturnValue: ReturnType<typeof useFiles> = {
        files: [],
        status: Status.Initial,
        getFiles: (() => undefined) as any,
        deleteFile: () => undefined,
        addFile: () => undefined,
        search: "foo",
        setSearch: () => undefined,
        message: "bar",
    };

    it("renders header", () => {
        mockedUseFiles.mockReturnValue(defaultUseFilesMockReturnValue);

        const subject = shallow(<App/>);

        expect(subject.find(Header)).toMatchSnapshot();
    });

    it("gets files on mount", () => {
        const getFiles: any = jest.fn();

        mockedUseFiles.mockReturnValue({
            ...defaultUseFilesMockReturnValue,
            getFiles,
        });

        shallow(<App/>);

        expect(getFiles).toHaveBeenLastCalledWith("");
    });

    it("renders loading state", () => {
        mockedUseFiles.mockReturnValue({
            ...defaultUseFilesMockReturnValue,
            status: Status.Loading,
        });

        const subject = shallow(<App/>);

        expect(subject.find("[data-loading]")).toMatchSnapshot();
        expect(subject.find("[data-error]")).toHaveLength(0);
        expect(subject.find(FileList)).toHaveLength(0);
    });

    it("renders error state", () => {
        mockedUseFiles.mockReturnValue({
            ...defaultUseFilesMockReturnValue,
            status: Status.Error,
            message: "baz",
        });

        const subject = shallow(<App/>);

        expect(subject.find("[data-loading]")).toHaveLength(0);
        expect(subject.find("[data-error]")).toMatchSnapshot();
        expect(subject.find(FileList)).toHaveLength(0);
    });

    it("renders error state", () => {
        const deleteFile = () => undefined;
        const files: FileInfo[] = [];

        mockedUseFiles.mockReturnValue({
            ...defaultUseFilesMockReturnValue,
            status: Status.Success,
            files,
            deleteFile,
        });

        const subject = shallow(<App/>);

        expect(subject.find("[data-loading]")).toHaveLength(0);
        expect(subject.find("[data-error]")).toHaveLength(0);
        expect(subject.find(FileList)).toMatchSnapshot();
    });
});