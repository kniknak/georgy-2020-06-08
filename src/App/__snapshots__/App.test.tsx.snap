// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders error state 1`] = `
<div
  data-error={true}
>
  Unexpected Error. 
  baz
</div>
`;

exports[`App renders error state 2`] = `
<FileList
  deleteFile={[Function]}
  files={Array []}
/>
`;

exports[`App renders header 1`] = `
<Header
  addFile={[Function]}
  search="foo"
  setSearch={[Function]}
/>
`;

exports[`App renders loading state 1`] = `
<div
  data-loading={true}
>
  Loading...
</div>
`;
